File: app/src/main/java/com/example/service/sip/SipEngine.kt
package com.example.service.sip

import android.content.Context
import android.util.Log
import com.example.data.model.SipConfig
import com.example.util.PhoneNumberSanitizer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import org.linphone.core.AVPFMode
import org.linphone.core.Account
import org.linphone.core.Call
import org.linphone.core.Core
import org.linphone.core.CoreListenerStub
import org.linphone.core.Factory
import org.linphone.core.LogLevel
import org.linphone.core.MediaEncryption
import org.linphone.core.RegistrationState
import org.linphone.core.TransportType

data class SipTestResult(
    val isSuccess: Boolean,
    val statusCode: Int,
    val message: String,
    val latencyMs: Long,
    val serverBanner: String = ""
)

interface SipCallEventListener {
    fun onConnecting(details: String)
    fun onRegistering(host: String)
    fun onRegistered(username: String)
    fun onRinging(remoteTag: String?)
    fun onConnected(audioCodec: String, localRtpPort: Int, remoteRtpPort: Int)
    fun onError(errorCode: Int, message: String)
    fun onEnded(reason: String)
    fun onAudioStatsUpdated(latencyMs: Int, packetsSent: Long, packetsReceived: Long)
}

/**
 * Single Linphone stack for REGISTER + INVITE.
 * Custom UDP registrar ([SipRegistrationEngine]) is isolated and is not the source of truth.
 *
 * Media: SRTP is offered and preferred. TLS transport is not forced because typical
 * UDP:5060 trunks (including user-entered sip.sipup.org) will fail a TLS handshake.
 * UI must only claim encryption when [isCurrentCallEncrypted] is true.
 */
class SipEngine private constructor(
    private val context: Context
) {
    private val tag = "SipEngine"
    private val engineJob = SupervisorJob()
    private val engineScope = CoroutineScope(Dispatchers.Main + engineJob)
    private var iterateJob: Job? = null
    private var core: Core? = null
    private var currentCall: Call? = null
    private var listener: SipCallEventListener? = null

    private var pendingCallDestination: String? = null
    private var pendingCallConfig: SipConfig? = null
    private var pendingCallCodec: G711CodecType? = null
    private var currentSipConfig: SipConfig? = null
    private var lastRegisterFingerprint: String? = null

    private val _registrationState = MutableStateFlow(SipRegistrationState())
    val registrationState: StateFlow<SipRegistrationState> = _registrationState.asStateFlow()

    val isRegistered: Boolean
        get() = _registrationState.value.isRegistered

    fun isCurrentCallEncrypted(): Boolean {
        val encryption = currentCall?.currentParams?.mediaEncryption ?: return false
        return encryption != MediaEncryption.None
    }

    fun setListener(listener: SipCallEventListener?) {
        this.listener = listener
    }

    private fun executePendingCall() {
        val dest = pendingCallDestination ?: return
        val config = pendingCallConfig ?: return
        pendingCallDestination = null

        val sanitized = PhoneNumberSanitizer.sanitizeDestination(dest)
        if (sanitized == null) {
            listener?.onError(400, "Invalid destination number")
            return
        }

        val c = core ?: return
        val callAddress = c.createAddress("sip:$sanitized@${config.host}")
        val callParams = c.createCallParams(null)

        val desiredIdentity = "sip:${config.username}@${config.host}"
        val account = c.accountList.find { it.params?.identityAddress?.asStringUriOnly() == desiredIdentity } ?: c.defaultAccount
        callParams?.account = account

        callParams?.setAudioEnabled(true)
        callParams?.setVideoEnabled(false)
        // Offer SRTP; do not require it so UDP trunks that lack SRTP still connect.
        callParams?.setMediaEncryption(MediaEncryption.SRTP)
        callParams?.setAvpfEnabled(false)
        callParams?.setCapabilityNegotiationsEnabled(false)
        callParams?.setCapabilityNegotiationReinviteEnabled(false)

        if (callAddress != null && callParams != null) {
            currentCall = c.inviteAddressWithParams(callAddress, callParams)
        } else {
            listener?.onError(500, "Failed to create call parameters")
        }
    }

    companion object {
        private val _lastSdpDiagnosticDump = MutableStateFlow<SdpDiagnosticDump?>(null)
        val lastSdpDiagnosticDump: StateFlow<SdpDiagnosticDump?> = _lastSdpDiagnosticDump

        @Volatile
        private var INSTANCE: SipEngine? = null

        fun getInstance(context: Context): SipEngine {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: SipEngine(context.applicationContext).also { INSTANCE = it }
            }
        }
    }

    private val coreListener = object : CoreListenerStub() {
        override fun onCallStateChanged(
            core: Core,
            call: Call,
            state: Call.State,
            message: String
        ) {
            Log.i(tag, "Call state changed: $state - $message")
            when (state) {
                Call.State.OutgoingProgress -> listener?.onConnecting(message)
                Call.State.OutgoingRinging -> listener?.onRinging(null)
                Call.State.Connected -> {
                    val codec = call.currentParams?.usedAudioPayloadType?.mimeType ?: "Unknown"
                    listener?.onConnected(codec, 0, 0)
                }
                Call.State.Error -> listener?.onError(500, message)
                Call.State.End, Call.State.Released -> listener?.onEnded(message)
                else -> {}
            }
        }

        override fun onAccountRegistrationStateChanged(
            core: Core,
            account: Account,
            state: RegistrationState,
            message: String
        ) {
            Log.i(tag, "Registration state: $state")
            val username = account.params?.identityAddress?.username ?: currentSipConfig?.username.orEmpty()
            val host = account.params?.serverAddress?.domain ?: currentSipConfig?.host.orEmpty()
            when (state) {
                RegistrationState.Progress -> {
                    _registrationState.value = _registrationState.value.copy(
                        status = RegistrationStatus.REGISTERING,
                        username = username,
                        host = host,
                        port = currentSipConfig?.port ?: 5060,
                        statusMessage = "Registering with $host...",
                        lastError = null
                    )
                    listener?.onRegistering(host)
                }
                RegistrationState.Ok -> {
                    _registrationState.value = _registrationState.value.copy(
                        status = RegistrationStatus.REGISTERED,
                        username = username,
                        host = host,
                        port = currentSipConfig?.port ?: 5060,
                        statusCode = 200,
                        statusMessage = "Registered with $host",
                        lastError = null,
                        serverBanner = "Linphone",
                        isKeepAliveActive = true
                    )
                    listener?.onRegistered(username)
                    executePendingCall()
                }
                RegistrationState.Failed -> {
                    _registrationState.value = _registrationState.value.copy(
                        status = RegistrationStatus.FAILED,
                        username = username,
                        host = host,
                        statusCode = 401,
                        statusMessage = message.ifBlank { "Registration failed" },
                        lastError = message,
                        isKeepAliveActive = false
                    )
                    listener?.onError(401, message)
                }
                RegistrationState.Cleared, RegistrationState.None -> {
                    if (_registrationState.value.status != RegistrationStatus.FAILED) {
                        _registrationState.value = _registrationState.value.copy(
                            status = RegistrationStatus.UNREGISTERED,
                            statusMessage = "Unregistered",
                            isKeepAliveActive = false
                        )
                    }
                }
                else -> {}
            }
        }
    }

    init {
        try {
            val factory = Factory.instance()

            val loggingService = factory.loggingService
            loggingService?.domain = "Linphone"
            loggingService?.setLogLevel(LogLevel.Message)

            core = factory.createCore(null, null, context)
            core?.setUserAgent("LinphoneAndroid", "5.2.0")

            val config = core?.config
            config?.setInt("rtp", "rtcp_xr_enabled", 0)
            config?.setInt("rtp", "rtcp_fb_implicit_rtcp_sync", 0)
            config?.setInt("sip", "use_rtcp", 0)
            config?.setInt("sip", "use_avpf", 0)
            config?.setInt("video", "capture", 0)
            config?.setInt("video", "display", 0)
            config?.setInt("video", "enabled", 0)

            val vp = core?.videoActivationPolicy
            vp?.setAutomaticallyAccept(false)
            vp?.setAutomaticallyInitiate(false)

            core?.setVideoCaptureEnabled(false)
            core?.setVideoDisplayEnabled(false)

            core?.audioPayloadTypes?.forEach { pt ->
                val mime = pt.mimeType.uppercase()
                if (mime == "PCMU" || mime == "PCMA" || mime == "OPUS" || mime.contains("TELEPHONE-EVENT")) {
                    pt.enable(true)
                } else {
                    pt.enable(false)
                }
            }

            // Prefer SRTP when the peer supports it. Do not mandate it: many UDP
            // registrars (user-configured sip.sipup.org:5060) have no TLS listener
            // and will reject a mandatory-SRTP / TLS-only INVITE.
            core?.setMediaEncryption(MediaEncryption.SRTP)
            core?.setMediaEncryptionMandatory(false)

            core?.addListener(coreListener)
            core?.start()
            core?.setNetworkReachable(true)

            iterateJob = engineScope.launch {
                while (isActive) {
                    core?.iterate()
                    delay(20)
                }
            }
        } catch (e: Exception) {
            Log.e(tag, "Failed to initialize Linphone Core", e)
        }
    }

    /**
     * Registers with Linphone. Cancels in-flight account setup first.
     * No-ops when host/user/password are unchanged.
     */
    fun register(sipConfig: SipConfig, force: Boolean = false) {
        if (!sipConfig.hasUsableCredentials()) {
            lastRegisterFingerprint = null
            currentSipConfig = null
            _registrationState.value = SipRegistrationState(
                status = RegistrationStatus.UNREGISTERED,
                statusMessage = "SIP credentials not configured",
                lastError = "Enter host, username, and password in Settings"
            )
            return
        }

        val fingerprint = sipConfig.registrationFingerprint()
        if (!force && fingerprint == lastRegisterFingerprint && _registrationState.value.isRegistered) {
            return
        }

        lastRegisterFingerprint = fingerprint
        currentSipConfig = sipConfig
        applyAccount(sipConfig, enableRegister = true)
    }

    fun refreshNow() {
        val config = currentSipConfig
        if (config == null || !config.hasUsableCredentials()) {
            _registrationState.value = _registrationState.value.copy(
                status = RegistrationStatus.UNREGISTERED,
                statusMessage = "SIP credentials not configured"
            )
            return
        }
        val account = core?.defaultAccount
        if (account != null) {
            core?.refreshRegisters()
        } else {
            register(config, force = true)
        }
    }

    fun unregister() {
        lastRegisterFingerprint = null
        currentSipConfig = null
        pendingCallDestination = null
        try {
            core?.clearAccounts()
            core?.clearProxyConfig()
            core?.clearAllAuthInfo()
        } catch (e: Exception) {
            Log.d(tag, "Unregister notice: ${e.message}")
        }
        _registrationState.value = SipRegistrationState(
            status = RegistrationStatus.UNREGISTERED,
            statusMessage = "Unregistered"
        )
    }

    fun startOutboundCall(
        sipConfig: SipConfig,
        destination: String,
        preferredCodec: G711CodecType?
    ) {
        val sanitized = PhoneNumberSanitizer.sanitizeDestination(destination)
        if (sanitized == null) {
            listener?.onError(400, "Invalid destination number")
            return
        }
        if (!sipConfig.hasUsableCredentials()) {
            listener?.onError(401, "SIP credentials not configured")
            return
        }

        val c = core ?: return
        applyAccount(sipConfig, enableRegister = true)
        lastRegisterFingerprint = sipConfig.registrationFingerprint()
        currentSipConfig = sipConfig

        c.setIpv6Enabled(false)
        c.setUseRfc2833ForDtmf(true)
        c.setUseInfoForDtmf(false)
        c.setAvpfMode(AVPFMode.Disabled)
        c.setMediaEncryption(MediaEncryption.SRTP)

        val targetMime = when (preferredCodec) {
            G711CodecType.PCMA -> "PCMA"
            G711CodecType.PCMU -> "PCMU"
            else -> "PCMU"
        }

        c.audioPayloadTypes?.forEach { pt ->
            val mime = pt.mimeType.uppercase()
            if (mime == targetMime || mime.contains("TELEPHONE-EVENT")) {
                pt.enable(true)
            } else {
                pt.enable(false)
            }
        }

        c.videoPayloadTypes?.forEach { pt ->
            pt.enable(false)
        }

        val natPolicy = c.createNatPolicy()
        if (natPolicy != null) {
            natPolicy.setIceEnabled(false)
            natPolicy.setStunEnabled(false)
            natPolicy.setTurnEnabled(false)
            c.setNatPolicy(natPolicy)
        }

        pendingCallDestination = sanitized
        pendingCallConfig = sipConfig
        pendingCallCodec = preferredCodec

        val desiredIdentity = "sip:${sipConfig.username}@${sipConfig.host}"
        val account = c.accountList.find { it.params?.identityAddress?.asStringUriOnly() == desiredIdentity } ?: c.defaultAccount
        if (account != null && account.state == RegistrationState.Ok) {
            executePendingCall()
        } else if (account == null) {
            listener?.onError(500, "Failed to create SIP account")
        }
    }

    suspend fun testSipConnection(sipConfig: SipConfig): SipTestResult = withContext(Dispatchers.IO) {
        if (!sipConfig.hasUsableCredentials()) {
            return@withContext SipTestResult(false, 0, "Host, username, and password are required", 0)
        }
        val start = System.currentTimeMillis()
        withContext(Dispatchers.Main) {
            register(sipConfig, force = true)
        }
        val observed = withTimeoutOrNull(12_000) {
            registrationState.first {
                it.status == RegistrationStatus.REGISTERED || it.status == RegistrationStatus.FAILED
            }
        }
        val elapsed = System.currentTimeMillis() - start
        when (observed?.status) {
            RegistrationStatus.REGISTERED -> SipTestResult(
                isSuccess = true,
                statusCode = observed.statusCode.takeIf { it > 0 } ?: 200,
                message = observed.statusMessage.ifBlank { "REGISTER succeeded" },
                latencyMs = elapsed,
                serverBanner = observed.serverBanner.ifBlank { "Linphone" }
            )
            RegistrationStatus.FAILED -> SipTestResult(
                isSuccess = false,
                statusCode = observed.statusCode.takeIf { it > 0 } ?: 401,
                message = observed.lastError ?: observed.statusMessage,
                latencyMs = elapsed,
                serverBanner = observed.serverBanner
            )
            else -> SipTestResult(false, 408, "REGISTER timed out", elapsed)
        }
    }

    fun stopCall(reason: String = "Call Ended") {
        currentCall?.terminate()
        currentCall = null
    }

    fun sendDtmfTone(digit: Char) {
        currentCall?.sendDtmf(digit)
    }

    fun updateDiagnosticDump(dump: SdpDiagnosticDump) {
        _lastSdpDiagnosticDump.value = dump
    }

    fun destroy() {
        iterateJob?.cancel()
        iterateJob = null
        try {
            currentCall?.terminate()
            core?.stop()
        } catch (e: Exception) {
            Log.d(tag, "Core stop notice: ${e.message}")
        }
        core = null
        engineJob.cancel()
        synchronized(SipEngine::class.java) {
            if (INSTANCE === this) {
                INSTANCE = null
            }
        }
    }

    private fun applyAccount(sipConfig: SipConfig, enableRegister: Boolean) {
        val c = core ?: return
        c.clearAccounts()
        c.clearProxyConfig()
        c.clearAllAuthInfo()

        val authInfo = Factory.instance().createAuthInfo(
            sipConfig.username,
            null,
            sipConfig.password,
            null,
            null,
            sipConfig.host
        )
        c.addAuthInfo(authInfo)

        val accountParams = c.createAccountParams() ?: return
        val identity = Factory.instance().createAddress("sip:${sipConfig.username}@${sipConfig.host}")
        accountParams.setIdentityAddress(identity)

        val serverAddr = Factory.instance().createAddress("sip:${sipConfig.host}:${sipConfig.port}")
        // UDP:5060 is the common registrar transport. Forcing TLS here would break
        // existing sip.sipup.org-style trunks that do not listen on 5061.
        serverAddr?.transport = TransportType.Udp
        accountParams.setServerAddress(serverAddr)
        accountParams.setOutboundProxyEnabled(true)
        accountParams.setRegisterEnabled(enableRegister)

        accountParams.setAvpfMode(AVPFMode.Disabled)
        accountParams.setRtpBundleEnabled(false)
        accountParams.setQualityReportingEnabled(false)
        accountParams.setPushNotificationAllowed(false)
        accountParams.setRemotePushNotificationAllowed(false)

        val natPolicy = c.createNatPolicy()
        if (natPolicy != null) {
            natPolicy.setIceEnabled(false)
            natPolicy.setStunEnabled(false)
            natPolicy.setTurnEnabled(false)
            accountParams.setNatPolicy(natPolicy)
        }

        val account = c.createAccount(accountParams)
        c.addAccount(account)
        c.defaultAccount = account

        if (enableRegister) {
            _registrationState.value = _registrationState.value.copy(
                status = RegistrationStatus.REGISTERING,
                username = sipConfig.username,
                host = sipConfig.host,
                port = sipConfig.port,
                statusMessage = "Registering with ${sipConfig.host}...",
                lastError = null
            )
        }
    }
}

File: ios/DialerID/Services/SipEngine.swift
import Combine
import Foundation

#if canImport(linphonesw)
import linphonesw
#endif

@MainActor
final class SipEngine: ObservableObject {
    static let shared = SipEngine()

    @Published private(set) var registrationState = SipRegistrationState()
    @Published private(set) var isRegistered = false
    @Published private(set) var hasActiveCall = false

    var onInitializing: ((String) -> Void)?
    var onDialing: ((String) -> Void)?
    var onConnecting: ((String) -> Void)?
    var onRinging: ((String?) -> Void)?
    var onEarlyMedia: ((String) -> Void)?
    var onConnected: ((String) -> Void)?
    var onError: ((Int, String) -> Void)?
    var onEnded: ((String) -> Void)?

    private var currentConfig: SipConfig?
    #if canImport(linphonesw)
    private var core: Core?
    private var currentCall: Call?
    private var iterateTimer: Timer?
    private var coreListener: CoreDelegateStub?
    private var triedAlternateTransport = false
    #endif

    func register(sipConfig: SipConfig, force: Bool = false) {
        let sipConfig = sipConfig.resolvedForRegistration()
        if sipConfig.needsPassword() {
            currentConfig = sipConfig
            registrationState = SipRegistrationState(
                status: .failed,
                username: sipConfig.username,
                host: sipConfig.host,
                port: sipConfig.port,
                statusMessage: "SIP password required",
                needsPassword: true
            )
            isRegistered = false
            return
        }
        if !sipConfig.hasUsableCredentials() {
            registrationState = SipRegistrationState(
                status: .unregistered,
                statusMessage: "SIP credentials not configured"
            )
            isRegistered = false
            return
        }
        currentConfig = sipConfig
        registrationState = SipRegistrationState(
            status: .registering,
            username: sipConfig.username,
            host: sipConfig.host,
            port: sipConfig.port,
            statusMessage: "Registering..."
        )
        #if canImport(linphonesw)
        triedAlternateTransport = false
        ensureCore()
        applyAccount(sipConfig, transport: .Udp)
        #else
        registrationState.status = .failed
        registrationState.lastError = "Linphone SDK is not linked. Run pod install on a Mac."
        registrationState.statusMessage = "Linphone SDK missing"
        #endif
    }

    func refreshNow(force: Bool = false) {
        guard let config = currentConfig else { return }
        register(sipConfig: config, force: true)
    }

    func unregister() {
        currentConfig = nil
        isRegistered = false
        hasActiveCall = false
        registrationState = SipRegistrationState(status: .unregistered, statusMessage: "Unregistered")
        #if canImport(linphonesw)
        try? core?.clearAccounts()
        try? core?.clearAllAuthInfo()
        #endif
    }

    func startOutboundCall(sipConfig: SipConfig, destination: String, preferredCodec: G711CodecType?) {
        let sipConfig = sipConfig.resolvedForRegistration()
        #if canImport(linphonesw)
        if currentCall != nil { return }
        #endif
        guard let sanitized = PhoneNumberSanitizer.sanitizeDestination(destination) else {
            onError?(400, "Invalid destination number")
            return
        }
        if sipConfig.needsPassword() {
            onError?(0, "SIP password required")
            return
        }
        if !sipConfig.hasUsableCredentials() {
            onError?(0, "SIP credentials not configured")
            return
        }
        register(sipConfig: sipConfig)
        onInitializing?("Initializing...")
        #if canImport(linphonesw)
        ensureCore()
        guard let core else {
            onError?(500, "SIP engine failed to start")
            return
        }
        let address = try? Factory.Instance.createAddress(addr: "sip:\(sanitized)@\(sipConfig.host)")
        guard let address else {
            onError?(400, "Invalid destination number")
            return
        }
        do {
            let params = try core.createCallParams(call: nil)
            params.mediaEncryption = .SRTP
            params.avpfMode = .Disabled
            currentCall = try core.inviteAddressWithParams(addr: address, params: params)
            hasActiveCall = currentCall != nil
            onDialing?("Dialing...")
        } catch {
            onError?(500, error.localizedDescription)
        }
        #else
        onError?(500, "Linphone SDK is not linked. Run pod install on a Mac.")
        #endif
    }

    func stopCall(reason: String = "Call Ended") {
        #if canImport(linphonesw)
        try? currentCall?.terminate()
        currentCall = nil
        #endif
        hasActiveCall = false
        onEnded?(reason)
    }

    func sendDtmf(_ digit: Character) {
        #if canImport(linphonesw)
        try? currentCall?.sendDtmf(dtmf: digit)
        #endif
    }

    func setMicrophoneMuted(_ muted: Bool) {
        #if canImport(linphonesw)
        core?.micEnabled = !muted
        #endif
    }

    func testSipConnection(_ sipConfig: SipConfig) async -> SipTestResult {
        let start = Date()
        let sipConfig = sipConfig.resolvedForRegistration()
        if sipConfig.needsPassword() {
            return SipTestResult(isSuccess: false, statusCode: 0, message: "SIP password required", latencyMs: 0)
        }
        if !sipConfig.hasUsableCredentials() {
            return SipTestResult(isSuccess: false, statusCode: 0, message: "Host, username, and password are required", latencyMs: 0)
        }
        register(sipConfig: sipConfig, force: true)
        let deadline = Date().addingTimeInterval(20)
        while Date() < deadline {
            if registrationState.status == .registered {
                let ms = Int64(Date().timeIntervalSince(start) * 1000)
                return SipTestResult(isSuccess: true, statusCode: 200, message: "REGISTER succeeded", latencyMs: ms, serverBanner: "Linphone")
            }
            if registrationState.status == .failed {
                let ms = Int64(Date().timeIntervalSince(start) * 1000)
                return SipTestResult(
                    isSuccess: false,
                    statusCode: registrationState.statusCode,
                    message: registrationState.lastError ?? registrationState.statusMessage,
                    latencyMs: ms
                )
            }
            try? await Task.sleep(nanoseconds: 200_000_000)
        }
        return SipTestResult(isSuccess: false, statusCode: 408, message: "REGISTER timed out", latencyMs: 20_000)
    }

    #if canImport(linphonesw)
    private func ensureCore() {
        if core != nil { return }
        do {
            let created = try Factory.Instance.createCore(configPath: nil, factoryConfigPath: nil, systemContext: nil)
            created.mediaEncryptionMandatory = false
            let listener = CoreDelegateStub(
                onAccountRegistrationStateChanged: { [weak self] _, _, state, message in
                    Task { @MainActor in
                        self?.handleRegistration(state: state, message: message)
                    }
                },
                onCallStateChanged: { [weak self] _, call, state, message in
                    Task { @MainActor in
                        self?.handleCall(call: call, state: state, message: message)
                    }
                }
            )
            created.addDelegate(delegate: listener)
            try created.start()
            core = created
            coreListener = listener
            iterateTimer = Timer.scheduledTimer(withTimeInterval: 0.02, repeats: true) { [weak created] _ in
                created?.iterate()
            }
        } catch {
            registrationState.status = .failed
            registrationState.lastError = error.localizedDescription
        }
    }

    private func applyAccount(_ sipConfig: SipConfig, transport: TransportType) {
        guard let core else { return }
        try? core.clearAccounts()
        try? core.clearAllAuthInfo()
        let identity = try? Factory.Instance.createAddress(addr: "sip:\(sipConfig.username)@\(sipConfig.host)")
        let server = try? Factory.Instance.createAddress(addr: "sip:\(sipConfig.host):\(sipConfig.port)")
        server?.transport = transport
        guard let identity, let server else { return }
        let auth = try? Factory.Instance.createAuthInfo(
            username: sipConfig.username,
            userid: sipConfig.username,
            passwd: sipConfig.password,
            ha1: nil,
            realm: nil,
            domain: sipConfig.host
        )
        if let auth {
            core.addAuthInfo(info: auth)
        }
        if let params = try? core.createAccountParams() {
            params.identityAddress = identity
            params.serverAddress = server
            params.registerEnabled = true
            params.expires = 300
            if let account = try? core.createAccount(params: params) {
                try? core.addAccount(account: account)
                core.defaultAccount = account
            }
        }
        core.stunServer = "stun.linphone.org"
    }

    private func handleRegistration(state: RegistrationState, message: String) {
        switch state {
        case .Progress:
            registrationState.status = .registering
            registrationState.statusMessage = message
            isRegistered = false
        case .Ok:
            registrationState.status = .registered
            registrationState.statusCode = 200
            registrationState.statusMessage = message
            isRegistered = true
        case .Failed:
            if !triedAlternateTransport, let config = currentConfig {
                triedAlternateTransport = true
                registrationState.status = .retrying
                registrationState.statusMessage = "Retrying over TCP..."
                applyAccount(config, transport: .Tcp)
                return
            }
            registrationState.status = .failed
            registrationState.statusMessage = message
            registrationState.lastError = message
            isRegistered = false
        case .Cleared:
            registrationState.status = .unregistered
            isRegistered = false
        default:
            break
        }
    }

    private func handleCall(call: Call, state: Call.State, message: String) {
        currentCall = call
        hasActiveCall = true
        switch state {
        case .OutgoingInit:
            onInitializing?(message)
        case .OutgoingProgress:
            onDialing?(message)
        case .OutgoingRinging:
            onRinging?(nil)
        case .OutgoingEarlyMedia:
            onEarlyMedia?(message)
        case .Connected, .StreamsRunning:
            onConnected?(call.currentParams?.usedAudioPayloadType?.mimeType ?? "G.711")
        case .Error:
            hasActiveCall = false
            currentCall = nil
            onError?(Int(call.reason.rawValue), message)
        case .End, .Released:
            hasActiveCall = false
            currentCall = nil
            onEnded?(message)
        default:
            break
        }
    }
    #endif
}
